File: backend/src/main/java/com/lsoft/aguapotable/service/PagoService.java
package com.lsoft.aguapotable.service;

import com.lsoft.aguapotable.domain.entity.Factura;
import com.lsoft.aguapotable.domain.entity.Pago;
import com.lsoft.aguapotable.domain.entity.Usuario;
import com.lsoft.aguapotable.domain.enums.EstadoFactura;
import com.lsoft.aguapotable.domain.enums.MetodoPago;
import com.lsoft.aguapotable.domain.repository.FacturaRepository;
import com.lsoft.aguapotable.domain.repository.PagoRepository;
import com.lsoft.aguapotable.exception.BusinessException;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;

@Service
@RequiredArgsConstructor
public class PagoService {

    private final PagoRepository pagoRepository;
    private final FacturaRepository facturaRepository;

    @Transactional
    public Pago registrarPago(Long facturaId, BigDecimal monto,
                               MetodoPago metodoPago, String referencia,
                               Usuario cajero) {
        Factura factura = facturaRepository.findById(facturaId)
                .orElseThrow(() -> new ResourceNotFoundException("Factura", facturaId));

        if (factura.getEstado() == EstadoFactura.PAGADA || factura.getEstado() == EstadoFactura.ANULADA) {
            throw new BusinessException("La factura " + factura.getNumeroFactura() + " no admite más pagos");
        }

        if (monto.compareTo(BigDecimal.ZERO) <= 0) {
            throw new BusinessException("El monto del pago debe ser mayor a cero");
        }

        Pago pago = Pago.builder()
                .factura(factura)
                .monto(monto)
                .fechaPago(LocalDate.now())
                .metodoPago(metodoPago)
                .referencia(referencia)
                .cajero(cajero)
                .build();

        pagoRepository.save(pago);
        actualizarEstadoFactura(factura);

        return pago;
    }

    private void actualizarEstadoFactura(Factura factura) {
        BigDecimal totalPagado = pagoRepository.sumPagadoByFactura(factura.getId());

        if (totalPagado.compareTo(factura.getTotalPagar()) >= 0) {
            factura.setEstado(EstadoFactura.PAGADA);
        } else if (totalPagado.compareTo(BigDecimal.ZERO) > 0) {
            factura.setEstado(EstadoFactura.PAGADA_PARCIAL);
        }

        facturaRepository.save(factura);
    }
}

File: backend/src/pago/pago.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Pago, Factura, Usuario } from '../common/entities';
import { EstadoFactura, MetodoPago } from '../common/enums';
import {
  BusinessException,
  ResourceNotFoundException,
} from '../common/exceptions';

@Injectable()
export class PagoService {
  constructor(
    @InjectRepository(Pago)
    private readonly pagoRepo: Repository<Pago>,
    @InjectRepository(Factura)
    private readonly facturaRepo: Repository<Factura>,
  ) {}

  async registrarPago(
    facturaId: number,
    monto: number,
    metodoPago: MetodoPago,
    referencia: string | undefined,
    cajero: Usuario | null,
  ): Promise<Pago> {
    const factura = await this.facturaRepo.findOne({
      where: { id: facturaId },
    });
    if (!factura)
      throw new ResourceNotFoundException('Factura', facturaId);

    if (
      factura.estado === EstadoFactura.PAGADA ||
      factura.estado === EstadoFactura.ANULADA
    ) {
      throw new BusinessException(
        'La factura ' + factura.numeroFactura + ' no admite más pagos',
      );
    }

    if (monto <= 0) {
      throw new BusinessException(
        'El monto del pago debe ser mayor a cero',
      );
    }

    const pago = this.pagoRepo.create({
      facturaId,
      monto: monto.toFixed(2),
      fechaPago: new Date().toISOString().split('T')[0],
      metodoPago,
      referencia: referencia ?? null,
      cajeroId: cajero?.id ?? null,
    });

    await this.pagoRepo.save(pago);
    await this.actualizarEstadoFactura(factura);

    return pago;
  }

  async getPorFactura(facturaId: number): Promise<Pago[]> {
    return this.pagoRepo.find({ where: { facturaId } });
  }

  private async actualizarEstadoFactura(factura: Factura): Promise<void> {
    const result = await this.pagoRepo
      .createQueryBuilder('p')
      .select('COALESCE(SUM(CAST(p.monto AS DECIMAL)), 0)', 'total')
      .where('p.factura_id = :facturaId', { facturaId: factura.id })
      .getRawOne();

    const totalPagado = parseFloat(result?.total ?? '0');
    const totalFactura = parseFloat(factura.totalPagar);

    if (totalPagado >= totalFactura) {
      factura.estado = EstadoFactura.PAGADA;
    } else if (totalPagado > 0) {
      factura.estado = EstadoFactura.PAGADA_PARCIAL;
    }

    await this.facturaRepo.save(factura);
  }
}
